senior/mbdesousa - sizefashion

mbdesousa opened the pull request.

vaporwavie commented:
Opa @mbdesousa, obrigado pelo input! Antes de iniciar a revisão, notei que os seus testes estão rodando e apontando erros. Consegue dar uma olhada nisso, por favor?

Valeu!

vaporwavie commented on sizefashion/tsconfig.json:
```diff
@@ -0,0 +1,58 @@
+      }
+    ],
+    "paths": {
+      "@components/*": [
+        "./app/components/*"
+      ],
+      "@context/*": [
+        "./app/context/*"
+      ],
+      "@interfaces/*": [
+        "./app/interfaces/*"
+      ],
+      "@service/*": [
+        "./app/service/*"
+      ],
+      "@assets/*": [
+        "./public/assets/*"
+      ],
+      "@/*": [
+        "./*"
```
@mbdesousa you declared the paths and everything but... you're not using them? :(

mbdesousa replied:
Haven't I?

vaporwavie reviewed (changes requested):
Thanks @mbdesousa! Here's a few considerations, along with my review:

- Make sure a linter is used during your development. It’s seen that on almost every component you’ve developed, you used single quotes instead of double quotes for elements property declarations. 
- React.FC is no longer recommended during component compositions. You can safely declare your props empirically. If your component does return a children from its parent, declare its types on your Props interface instead (e.g { children: React.ReactNode })
- For context custom hooks, you don’t need to prefix them with `useSomethingContext`. You can safely name it `useSomething`.
- Avoid naming your types and interfaces starting with a capital I. This is something very common for C# and C++ development, but it can get confusing for frontend development. If you have an interface that types the props for your component, and you component is properly isolated, an appropriate name would be `Props` instead of `IProps`. If you’re on a scenario where multiple components are part of the same file, you can go with `ComponentNameProps`.
- It’s cool that you bootstrapped your project with NextJS, but almost 85% of its power wasn’t even applied for the challenge. No `getServerSideProps`, no `getInitialProps`. The main reason to adopt NextJS on a frontend project is to have full SSR, SEO and dynamic progressive development covered by default. If you’re only going for the routing (which you didn’t entirely used it, given the pages aren’t SSRed), react-router-dom with vite would work much better, as well as requiring less code/boilerplate.
- Unit tests were part of the assignment, but I haven't seen any.
- You’ve committed both package-lock.json and yarn.lock. By having two completely different lock files, you enable the developer opening your code to have mismatching dependency signatures and potentially breaking your code composition. Given your install instructions only take in consideration `yarn`, I believe `package-lock.json` should be removed from your code.

mbdesousa commented:
Hey there @vaporwavie 

As you mentioned, I haven't utilized all the potential of Next.js, and I completely agree with you. I'm not very experienced with Next.js, so this project has been a great learning opportunity for me. I am currently addressing the issues you pointed out, and I'd like to share my thoughts on some of your points:

- Initially, my challenge arose when I was asked not to use what I'm most familiar with, which is CRA (Create React App). However, it's not a major problem since I had some experience with Next.js a few months ago, and its foundation in React gives me confidence in using it. I must admit that my prior experience mainly revolved around refactoring, maintaining, and updating old, complex codebases. As a result, I haven't had many opportunities to explore the initial stages of a project like this one. Specially with Next.
- Regarding the naming convention for Context, I usually prefer using the word "Context" to differentiate it from custom hooks. However, I am adaptable and open to using another convention. As developers, being flexible is essential.
- I realized that using so much "use client" in a server-side rendering framework like Next.js was not the best approach. Manipulating the DOM directly was not suitable for SSR; it was more of a way to demonstrate some knowledge I gained while working on a design system.
- Since I am more experienced with older versions of React, I haven't kept up with the latest updates as much as I should have. That's why I am very grateful for the documentation you provided. It will be immensely helpful.

I believe that covers about everything.

I will take into consideration all the points you've raised, and I'll start working on the improvements as soon as possible. You'll hear from me soon with the updates. Thank you for your guidance and support.

mbdesousa commented on sizefashion/app/components/Carousel/index.tsx:
```diff
@@ -0,0 +1,97 @@
+              style={{ backgroundColor: 'white', minWidth: 360 }}
+              key={key}
+            >
+              {/* YOLO */}<div dangerouslySetInnerHTML={{ __html: item.image }} />
```
@vaporwavie given that the payload appears as shown in the screenshot below, I opted for two different rendering approaches. However, there might be a more efficient way to handle this.

![image](https://github.com/sizebay/frontend-challenge/assets/99999956/e7621ac8-bee2-4f5b-bd2f-12188563f7e4)

vaporwavie replied:
@mbdesousa here's a cool solution to adopt with React, in two steps:

### 1. Sanitizing the string element
```js
// Given this dirty string
let dirtyProductImage =
 "<img src=\"//store.sizebay.com/cdn/shop/products/foto-05_large.png?v=1601062306\" alt=\"Nylon Bomber Jacket in Stone\" class=\"product__img\">";

// We first remove the backlashes
dirtyProductImage = dirtyProductImage.replace(/\"/g, `"`);

// Then we append the missing closing tag to prevent breaking layouts (while also increasing a11y coverage)
const storeLogo = `${dirtyProductImage}</img>`;
```

### 2. Rendering it
```jsx
function ProductImage({ productImage }) {
  const imageSlotRef = useRef(null);

  useEffect(() => {
    if (!imageSlotRef.current) return;

    // We're going to replace the image element with the one brought from props
    imageSlotRef.current.innerHTML = productImage;
  }, [imageSlotRef]);

  return (
    <div style={{ display: "contents" }} ref={imageSlotRef}>
      {/* We're going to replace this image element with the one brought from props */}
    </div>
  );
}

```

Using `dangerouslySetInnerHTML` can expose your application to XSS atacks. See at: https://legacy.reactjs.org/docs/dom-elements.html#dangerouslysetinnerhtml

vaporwavie reviewed (approved):
Thanks @mbdesousa! We'll provide you some final feedback over the challenge in a bit :)

Closing this challenge.
